### DISCUSSION TITLE
Optional path segment with regex?

### DISCUSSION THREAD

**Discussion by jmandel** [honojs/hono#1536] [Q&A]
I'd like to express something like: an optional "type" parameter can be supplied, if it starts with "c".

Things I've tried:

```ts
app.get('/api/animal/:type{c.+}?/show', (c) => c.text('Animal!' + JSON.stringify(c.req.param())))
// ^ does not match types when supplied

app.get('/api/animal/:type?{c.+}/show', (c) => c.text('Animal!' + JSON.stringify(c.req.param())))
// ^ Matches a required param called `type?`; does not match paths that omit the type segment

app.get('/api/animal/:type{(c.+)?}/show', (c) => c.text('Animal!' + JSON.stringify(c.req.param())))
// Internal Server Error
```

I get "Internal Server Error" for any regex involving `(`, which also prevents me from expressing a required path segment like `(a|b)`:

    app.get('/api/animal/:type{(a|b)}/show', (c) => c.text('Animal!' + JSON.stringify(c.req.param())))

Am I doing something wrong? Is it possible to express optional-regex-backed path segments? Is it possible to express required-regex-backed path segments that include parens in the regex?

**Comment by EdamAme-x** [comment]
I tried a similar case and got the following error.

The problem arises here.
[https://deno.land/x/hono@v3.7.3/router/reg-exp-router/router.ts?source=#L202](https://deno.land/x/hono@v3.7.3/router/reg-exp-router/router.ts?source=#L202)

```
TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
    at RegExpRouter.match (https://deno.land/x/hono@v3.7.3/router/reg-exp-router/router.ts:202:36)
    at Hono.matchRoute (https://deno.land/x/hono@v3.7.3/hono-base.ts:277:24)
    at Hono.dispatch (https://deno.land/x/hono@v3.7.3/hono-base.ts:301:39)
    at Server.fetch (https://deno.land/x/hono@v3.7.3/hono-base.ts:375:17)
    at Server.#respond (https://deno.land/std@0.203.0/http/server.ts:305:37)
    at Server.#serveHttp (https://deno.land/std@0.203.0/http/server.ts:352:20)
    at eventLoopTick (ext:core/01_core.js:183:11)
```

**Comment by EdamAme-x** [comment]
It works if.
`()` seems to be the cause.

```diff
import { Hono } from "https://deno.land/x/hono/mod.ts";
import { serve } from "https://deno.land/std/http/mod.ts";

const app = new Hono();

- app.get("/api/animal/:type{(c.+)?}/show", (c) =>
+ app.get("/api/animal/:type{c.+?}/show", (c) =>
  c.text("Animal!" + JSON.stringify(c.req.param()))
);

serve(app.fetch);
```

**Reply by jmandel** [reply]
Putting a question mark without parens changes the semantics though. It turns the + into a non-greedy operator, but that expression still needs a "c" + one other character, and won't let you skip the path segment altogether, right?

**Reply by EdamAme-x** [reply]
Yes.
This may be a bug on the HONO side, so this may be the only way... 

```js
import { Hono } from "https://deno.land/x/hono/mod.ts";
import { serve } from "https://deno.land/std/http/mod.ts";

const app = new Hono();

app.get("/api/animal/:type/show", (c) => {
  if (/(c.+)?/.test(c.req.param("type"))) {
    return c.text("Animal!" + JSON.stringify(c.req.param()));
  }
});

serve(app.fetch);
```

**Reply by raphaelkieling** [reply]
Is it occurring only on deno?
